packages/flutter/lib/src/rendering/selection.dart

```diff
@@ -375,6 +399,114 @@ class SelectionEdgeUpdateEvent extends SelectionEvent {
   final Offset globalPosition;
 }
 
+/// Extends the start or end of the selection by a given [TextGranularity].
+///
+/// To handle this event, move the associated selection edge, as dictated by
+/// [isEnd], according to the [granularity].
+class GranularlyExtendSelectionEvent extends SelectionEvent {
+  /// Creates a [GranularlyExtendSelectionEvent].
+  ///
+  /// All parameters are required and must not be null.
+  const GranularlyExtendSelectionEvent({
+    required this.forward,
+    required this.isEnd,
+    required this.granularity,
+  }) : super._(SelectionEventType.granularlyExtendSelection);
+
+  /// Whether to extend the selection forward.
+  final bool forward;
+
+  /// Whether this event is updating the end selection edge.
+  final bool isEnd;
+
+  /// The granularity for which the selection extend.
+  final TextGranularity granularity;
+}
+
+/// The directional to extend a selection.
+///
+/// The [DirectionallyExtendSelectionEvent] uses this enum to describe how
+/// [Selectable] should extend their selection.
+enum SelectionExtendDirection {
+  /// Move one edge of the selection vertically to the previous adjacent line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
+  previousLine,
+
+  /// Move one edge of the selection vertically to the next adjacent line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
+  nextLine,
+
+  /// Move the selection edges forward to a certain horizontal offset in the
+  /// same line.
+  ///
+  /// If there is no on-going selection, the selection must start with the first
+  /// line (or equivalence of first line in non-text selectable) and select
+  /// toward the horizontal offset in the same line.
+  ///
+  /// See [DirectionallyExtendSelectionEvent.dx] on how to
+  /// calculate the horizontal offset.
```

Comment:
What if the horizontal offset is extends beyond the width of the field, does it wrap to the next line or stop at this line?

Maybe here isn't the place to explain that, but just something I was wondering while reading these docs.

Reply:
I don't think we can define how to handle these cases here. It depends on the selectable they may choose to select things differently